General EnTAP Flags
============================

These are the general EnTAP flags that can be used during Execution. They are not specific to a certain stage of the pipeline. These will be used via the command line (denoted CMD) or ini file (denoted INI).

*-*-runP / *-*-runN [CMD]
-------------------------
* Specify a blastp or blastx annotation
* If - -runP is selected with a nucleotide input, frame selection will be ran and annotation stages will be executed with protein sequences (blastp)
* If - -runP is selected with a protein input, frame selection will not be ran and annotation will be executed with protein sequences (blastp)
* If - -runN is selected with nucleotide input, frame selection will not be ran and annotation will be executed with nucleotide sequences (blastx)

-i / *-*-input [string] [CMD]
----------------------------------
* Path to the transcriptome file (either nucleotide or protein)

-d / *-*-database [multi-string] [CMD]
-----------------------------------------------
    * Specify up to 5 DIAMOND indexed (.dmnd) databases to run similarity search against

*-*-|flag_path| [string] [CMD]
------------------------------------
* Point to |config_file| to specify file paths
* Default: |config_file| residing in the current working directory

*-*-out-dir [string] [CMD]
-------------------------------
* Specify an output directory for all of the files generated by EnTAP
* Default: Current working directory

-t / *-*-threads [integer] [CMD]
-------------------------------------
* Specify thread number for execution

*-*-out-dir [string] [CMD]
--------------------------------
* Specify an output directory for all of the files generated by EnTAP
* Default: Current working directory

*-*-overwrite [CMD]
-----------------------
    * All previously ran files will be overwritten
    * Without this flag EnTAP will recognize previous runs and skip portions that were already ran

*-*-graph [CMD]
-----------------
    * This will check whether or not your system has graphing functionality supported and exit
    * If Python with the Matplotlib module are installed on your system graphing should be enabled!
    * This can be specified on its own

-t / *-*-threads [integer]
--------------------------------
    * Specify the number of threads of execution

*-*-no-trim [CMD]
-----------------------
    * By default, EnTAP will trim your sequence headers to the first space to maintain compatbility across different software. Using this flag will instead retain the information of the header by removing all spaces.
    * Example: 
   
        * >TRINITY_231.1 protein12312_43 inform
        * >TRINITY_231.1protein12312_43inform

    * A word of caution when using this flag. EnTAP may have difficulty matching sequence headers from BAM/SAM files to your input transcriptome during Expression Analysis. You will receive an error if this occurs. 

*-*-state [string] [CMD]
-----------------------------
    * Precise control over execution :ref:`stages<state-label>`. This flag allows for certain parts to be ran while skipping others. 
    * Warning: This may cause issues depending on what you plan on running! 

*-*-version [CMD]
-------------------
    * Prints the current EnTAP version you are running

*-*-no-check [CMD]
---------------------
    * EnTAP checks execution paths and inputs prior to annotating to prevent finding out your input was wrong until midway through a run. Using this flag will eliminate the check (not advised to use!)

*-*-output-format [multi-integer] [INI]
--------------------------------------------
    * Specify multiple output file formats for each stage of the pipeline

        * 1. TSV File (default)
        * 2. CSV File
        * 3. FASTA Protein File (default)
        * 4. FASTA Nucleotide File (default)
        * 5. Gene Enrichment Gene ID + Effective Length
        * 6. Gene Enrichment Gene ID + GO Terms
        * 7. Gene Ontology Terms (Sequence ID,GO Term ID, GO Term, Category, and Sequence Effective Length) TSV format (default)